File: bindings/node/examples/custom-template.js
// Import the Ambi Agent and helpers.
const {Agent, LLMEngineConfig, deepseekTemplate} = require('../lib')

// ---- Custom chat template example ----
// Shows how to use a built-in template or define a fully custom one.
// Run:  OPENAI_API_KEY=sk-... node examples/custom-template.js

async function main() {
    const apiKey = process.env.OPENAI_API_KEY || 'sk-your-key-here'

    // Step 1: Configure the remote LLM engine.
    const engineConfig = LLMEngineConfig.openai({
        apiKey,
        baseUrl: 'https://api.openai.com/v1',
        modelName: 'gpt-4o-mini',
        temp: 0.7,
        topP: 0.9,
    })

    // --- Option A: Use a built-in template (returned as a JS object) ---
    const builtInTpl = deepseekTemplate()  // also: chatmlTemplate(), llama3Template(), qwenTemplate() …
    console.log('DeepSeek template system prefix:', JSON.stringify(builtInTpl.systemPrefix))

    // --- Option B: Define a fully custom template ---
    const customTpl = {
        systemPrefix: '<|SYS_START|>\n',
        systemSuffix: '\n<|SYS_END|>\n\n',
        userPrefix: '<|HUMAN|>: ',
        userSuffix: '\n',
        assistantPrefix: '<|BOT|>: ',
        assistantSuffix: '\n',
        thinkPrefix: '',
        thinkSuffix: '',
        toolPrefix: '<|TOOL_EXECUTION|>\n',
        toolSuffix: '\n<|END_EXECUTION|>\n',
        toolIdPrefix: '',
        toolIdSuffix: '',
        mediaPlaceholder: '',
    }

    // Step 2: Mount the custom template onto the Agent.
    const _agent = (await Agent.make(engineConfig)).customTemplate(customTpl)

    console.log('Agent created with custom template. Ready for chat.')
    // Use it like: const res = await runner.chat(agent, state, "Hello")
}

main().catch(console.error)

File: bindings/python/examples/custom-template.py
"""
custom-template.py  -- Custom chat template

Shows how to use a built-in template or define a fully custom one.
Usage:    python examples/custom-template.py
"""

import asyncio
import os
from ambi_python import Agent, LLMEngineConfig, deepseek_template


async def main():
    api_key = os.environ.get("OPENAI_API_KEY", "sk-your-key-here")

    engine_config = LLMEngineConfig.openai(
        api_key=api_key,
        base_url="https://api.openai.com/v1",
        model_name="gpt-4o-mini",
        temp=0.7,
        top_p=0.9,
    )

    # --- Option A: Use a built-in template ---
    built_in = deepseek_template()
    print("DeepSeek template system_prefix:", built_in["system_prefix"])

    # --- Option B: Define a fully custom template ---
    # All 13 template fields must be provided as keyword arguments.
    agent = await Agent.make(engine_config)
    agent = agent.custom_template(
        system_prefix="<|SYS_START|>\n",
        system_suffix="\n<|SYS_END|>\n\n",
        user_prefix="<|HUMAN|>: ",
        user_suffix="\n",
        assistant_prefix="<|BOT|>: ",
        assistant_suffix="\n",
        think_prefix="",
        think_suffix="",
        tool_prefix="<|TOOL_EXECUTION|>\n",
        tool_suffix="\n<|END_EXECUTION|>\n",
        tool_id_prefix="",
        tool_id_suffix="",
        media_placeholder="",
    )

    print("Agent created with custom template. Ready for chat.")


asyncio.run(main())
